File: services/alerts/alert_types/free_trial.ts
import { AlertRecord, EmailConfig } from "../index.ts";
import { Recipient } from "../template.ts";
import { commonTemplate } from "../template.ts";

interface FreeTrialArguments {
    tenant: string;
    // This feels like it should apply to all alert types, and doesn't belong here..
    recipients: Recipient[];
    trial_start: string;
    trial_end: string;
    has_credit_card: boolean;
}

type FreeTrialRecord = AlertRecord<"free_trial", FreeTrialArguments>;

const freeTrialStarted = (req: FreeTrialRecord): EmailConfig[] => {
    const args = req.arguments;

    return args.recipients.map((recipient) => ({
        emails: [recipient.email],
        subject: "Estuary Free Trial",
        content: args.has_credit_card
            // The only scenario in which your trial would start and you have a cc on file is we manually gave you back your trial
            ? commonTemplate(
                `
                <mj-text>Your Estuary account <a class="identifier">${args.tenant}</a> has started its 30-day free trial. This trial will end on <strong>${args.trial_end}</strong>. Billing will begin accruing then.</mj-text>
            `,
                recipient,
            )
            : commonTemplate(
                `
                <mj-text>We hope you're enjoying Estuary Flow. Our free tier includes 10 GB/month and 2 connectors. Your account <a class="identifier">${args.tenant}</a> has now exceeded that, so it has been transitioned to a 30 day free trial ending on <strong>${args.trial_end}</strong>.</mj-text>
                <mj-text>Please add payment information in the next 30 days to continue using the platform. If you have any questions, feel free to reach out to <a href="mailto:support@estuary.dev">support@estuary.dev</a></mj-text>
                <mj-button href="https://dashboard.estuary.dev/admin/billing">Add payment information</mj-button>
            `,
                recipient,
            ),
    }));
};

const freeTrialEnded = (req: FreeTrialRecord): EmailConfig[] => {
    const args = req.resolved_arguments;
    if (!args) {
        throw new Error("Resolved arguments are required for this alert type.");
    }

    return args.recipients.map((recipient) => ({
        emails: [recipient.email],
        subject: args.has_credit_card ? "Estuary Flow: Paid Tier" : "Estuary Paid Tier: Enter Payment Info to Continue Access",
        content: args.has_credit_card
            ? commonTemplate(
                `
                <mj-text>We hope you are enjoying Estuary Flow. Your free trial for account <a class="identifier">${args.tenant}</a> is officially over.</mj-text>
                <mj-text>Since you have already added a payment method, no action is required. If you have any questions, feel free to reach out to <a href="mailto:support@estuary.dev">support@estuary.dev</a> anytime!</mj-text>
                <mj-button href="https://dashboard.estuary.dev">🌊 View your data flows</mj-button>
                `,
                recipient,
            )
            : commonTemplate(
                `
                <mj-text>We hope you are enjoying Estuary Flow. Your free trial for account <a class="identifier">${args.tenant}</a> is officially over.</mj-text>
                <mj-text>Please add payment information immediately to continue using the platform. If you have any questions, feel free to reach out to <a href="mailto:support@estuary.dev">support@estuary.dev</a></mj-text>
                <mj-button href="https://dashboard.estuary.dev/admin/billing">Add payment information</mj-button>
            `,
                recipient,
            ),
    }));
};

export const freeTrialEmail = (request: FreeTrialRecord): EmailConfig[] => {
    if (request.resolved_at) {
        return freeTrialEnded(request);
    } else {
        return freeTrialStarted(request);
    }
};

File: crates/alert-notifications/src/free_trial.rs
use super::template_names;
use anyhow::Context;

pub fn register_templates<'a>(registry: &mut handlebars::Handlebars<'a>) -> anyhow::Result<()> {
    let (fired_subject, fired_body) = template_names(models::status::AlertType::FreeTrial, false);
    registry
        .register_template_string(&fired_subject, r#"Estuary Free Trial"#)
        .context("registering free_trial-fired-subject template")?;

    registry
        .register_template_string(
            &fired_body,
            r#"{{#if arguments.has_credit_card}}
<p class="body-text">Your Estuary account <span class="identifier">{{arguments.tenant}}</span> has started its 30-day free trial. This trial will end on <strong>{{arguments.trial_end}}</strong>. Billing will begin accruing then.</p>
{{else}}
<p class="body-text">We hope you're enjoying Estuary Flow. Our free tier includes 10 GB/month and 2 connectors. Your account <span class="identifier">{{arguments.tenant}}</span> has now exceeded that, so it has been transitioned to a 30 day free trial ending on <strong>{{arguments.trial_end}}</strong>.</p>
<p class="body-text">Please add payment information in the next 30 days to continue using the platform. If you have any questions, feel free to reach out to <a href="mailto:support@estuary.dev">support@estuary.dev</a></p>
<a href="{{> dashboard_billing_url}}" class="button">Add payment information</a>
{{/if}}"#,
        )
        .context("registering free_trial-fired-body template")?;

    // Resolved state templates
    let (resolved_subject, resolved_body) =
        template_names(models::status::AlertType::FreeTrial, true);
    registry
        .register_template_string(
            &resolved_subject,
            r#"{{#if arguments.has_credit_card}}Estuary Flow: Paid Tier{{else}}Estuary Paid Tier: Enter Payment Info to Continue Access{{/if}}"#,
        )
        .context("registering free_trial-resolved-subject template")?;

    registry
        .register_template_string(
            &resolved_body,
            r#"{{#if arguments.has_credit_card}}
<p class="body-text">We hope you are enjoying Estuary Flow. Your free trial for account <span class="identifier">{{arguments.tenant}}</span> is officially over.</p>
<p class="body-text">Since you have already added a payment method, no action is required. If you have any questions, feel free to reach out to <a href="mailto:support@estuary.dev">support@estuary.dev</a> anytime!</p>
<a href="https://dashboard.estuary.dev" class="button">🌊 View your data flows</a>
{{else}}
<p class="body-text">We hope you are enjoying Estuary Flow. Your free trial for account <span class="identifier">{{arguments.tenant}}</span> is officially over.</p>
<p class="body-text">Please add payment information immediately to continue using the platform. If you have any questions, feel free to reach out to <a href="mailto:support@estuary.dev">support@estuary.dev</a></p>
<a href="{{> dashboard_billing_url}}" class="button">Add payment information</a>
{{/if}}"#,
        )
        .context("registering free_trial-resolved-body template")?;

    Ok(())
}
